.github/workflows/docker-build-and-deploy.yml

```diff
@@ -0,0 +1,92 @@
+name: Docker Build and Deploy
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+  workflow_dispatch: # This allows you to manually trigger the workflow
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+      - name: use node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: '18.x'
+      - run: npm ci
+      - run: npm run test
+  build:
+    needs: test
+    runs-on: ubuntu-latest
+    environment: dev
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Extract Git commit hash
+        id: vars
+        run: echo "GIT_SHA=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
+
+      - name: Build Docker image with commit hash
+        run: docker build -t house-innovations-housnav-web:${{ env.GIT_SHA }} -f apps/web/Dockerfile . --no-cache
+
+      - name: Tag Docker image for PR
+        if: github.event_name == 'pull_request'
+        run: docker tag house-innovations-housnav-web:${{ env.GIT_SHA }} house-innovations-housnav-web:pr-${{ github.event.number }}
+
+      - name: Tag Docker image as latest for main branch
+        if: github.ref == 'refs/heads/main'
+        run: docker tag house-innovations-housnav-web:${{ env.GIT_SHA }} house-innovations-housnav-web:latest
+
+      - name: Login to OpenShift image registry with Docker
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ secrets.OPENSHIFT_REGISTRY_NAME }}
+          username: ${{ secrets.OPENSHIFT_SERVICEACCOUNT_NAME }}
+          password: ${{ secrets.OPENSHIFT_SERVICEACCOUNT_TOKEN }}
+
+      - name: Tag Docker image for OpenShift registry
+        run: docker tag house-innovations-housnav-web:${{ env.GIT_SHA }} ${{ secrets.OPENSHIFT_REGISTRY_REPOSITORY }}:${{ env.GIT_SHA }}
+
+      - name: Push Docker image to OpenShift registry
+        run: docker push ${{ secrets.OPENSHIFT_REGISTRY_REPOSITORY }}:${{ env.GIT_SHA }}
+
+  deploy:
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```

Comment:
Would we want this to all be in the same workflow or might we want to break out the PR tests and build into a separate workflow file?

That way it might be clearer what's happening on each and you wouldn't have to do this check.

Reply:
Working on it, I thought about it but didn't want to spend too much time optimizing since there has already been a lot of time on this.